[2/n] Consolidate `replicate` and `DDP`: split `forward` function

yhcharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96658

Split `forward` function into `pre_forward` and `post_forward`, so that they can be reused in the composable API of `replicate`.

Differential Revision: [D44377456](https://our.internmc.facebook.com/intern/diff/D44377456)

yhcharles committed:
[2/n] Consolidate `replicate` and `DDP`: split `forward` function
Split `forward` function into `pre_forward` and `post_forward`, so that they can be reused in the composable API of `replicate`.

[ghstack-poisoned]

yhcharles committed:
Update on "[2/n] Consolidate `replicate` and `DDP`: split `forward` function"


Split `forward` function into `pre_forward` and `post_forward`, so that they can be reused in the composable API of `replicate`.

[ghstack-poisoned]

yhcharles committed:
Update on "[2/n] Consolidate `replicate` and `DDP`: split `forward` function"


Split `forward` function into `pre_forward` and `post_forward`, so that they can be reused in the composable API of `replicate`.

[ghstack-poisoned]

yhcharles committed:
Update on "[2/n] Consolidate `replicate` and `DDP`: split `forward` function"


Split `forward` function into `pre_forward` and `post_forward`, so that they can be reused in the composable API of `replicate`.

[ghstack-poisoned]

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -1542,6 +1535,17 @@ def forward(self, *inputs, **kwargs):
 
+    def forward(self, *inputs, **kwargs):
+        with torch.autograd.profiler.record_function("DistributedDataParallel.forward"):
+            if self._delay_all_reduce_all_params:
```
why don't we move this logic to pre-forward?

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -1542,6 +1535,17 @@ def forward(self, *inputs, **kwargs):
+        with torch.autograd.profiler.record_function("DistributedDataParallel.forward"):
+            if self._delay_all_reduce_all_params:
+                output = self.module.forward(*inputs, **kwargs)
+                self._clear_grad_buffer()
```
why is this needed only for delay AR?

yhcharles replied:
Thanks for pointing out. I think I can move `self._clear_grad_buffer()` to `_post_forward()` as well. Let me update.
PS: The motivation is to move all operations before `self.module.forward()` into `_pre_forward()`, and operations after `self.module.forward()` into `_post_forward()`. So that `_pre_forward()` and `_post_forward()` can be reused in composable APIs.

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -2115,10 +2119,7 @@ def _check_comm_hook(self, hook):
-            "bf16_compress_hook",
-            "bf16_compress_wrapper_hook",
-        ] and (
+        if hook.__name__ in ["bf16_compress_hook", "bf16_compress_wrapper_hook"] and (
```
nit: unrelated change?

rohan-varma reviewed (commented):
just a small q, thanks!

yhcharles committed:
Update on "[2/n] Consolidate `replicate` and `DDP`: split `forward` function"


Split `forward` function into `pre_forward` and `post_forward`, so that they can be reused in the composable API of `replicate`.

[ghstack-poisoned]

yhcharles committed:
Update on "[2/n] Consolidate `replicate` and `DDP`: split `forward` function"


Split `forward` function into `pre_forward` and `post_forward`, so that they can be reused in the composable API of `replicate`.

[ghstack-poisoned]

rohan-varma reviewed (commented):
LGTM overall, since this is a large change, can we run our internal QPS benchmarks and confirm no regression? Thanks!

yhcharles commented:
@yhcharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44377456).

rohan-varma reviewed (approved):
looks good, thanks for running benchmark and helping the consolidation!

yhcharles commented:
Benchmark done. No regresison for QPS or accuracy. Results can be found in https://www.internalfb.com/diff/D44377456

yhcharles commented:
@pytorchbot merge --green

yhcharles commented:
@pytorchbot merge -g

yhcharles commented:
@pytorchbot merge

yhcharles commented:
@pytorchbot rebase

yhcharles committed:
Update on "[2/n] Consolidate `replicate` and `DDP`: split `forward` function"


Split `forward` function into `pre_forward` and `post_forward`, so that they can be reused in the composable API of `replicate`.

Differential Revision: [D44377456](https://our.internmc.facebook.com/intern/diff/D44377456)

[ghstack-poisoned]

yhcharles commented:
@pytorchbot merge

yhcharles commented:
@yhcharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44464781).

yhcharles commented:
@yhcharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44377456).
